---
title: "PBMC single cell RNA-editing workflow"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
```

```{r}
library(raer)
library(DropletUtils)
library(scuttle)
library(scran)
library(scater)
library(SingleCellExperiment)
library(BiocParallel)
data_dir <- path.expand("~/rbi/src/raerdata/")
```

## Process single cell data

Just following the basic Bioc pipeline from the OSCA ebook. 

```{r}
if(!file.exists(file.path(data_dir, "human-pbmc", "pbmc.rds"))){
  
  bpp <- MulticoreParam()
  
  if(!file.exists(file.path(data_dir, 
                            "human-pbmc",
                            "filtered_feature_bc_matrix", 
                            "matrix.mtx.gz"))){
    untar(file.path(data_dir, "human-pbmc",
                    "10k_PBMC_3p_nextgem_Chromium_X_filtered_feature_bc_matrix.tar.gz"),
          exdir = file.path(data_dir, 
                            "human-pbmc"))
  }
  sce <- read10xCounts(file.path(data_dir, 
                                 "human-pbmc",
                                 "filtered_feature_bc_matrix"), col.names=TRUE)
  
  rownames(sce) <- uniquifyFeatureNames(rowData(sce)$ID, rowData(sce)$Symbol)
  
  sce <- logNormCounts(sce)
  
  dec <- modelGeneVarByPoisson(sce, BPPARAM=bpp)
  top_var_genes <- getTopHVGs(dec, n = 2000)
  
  set.seed(10000)
  sce <- denoisePCA(sce, subset.row=top_var_genes, technical=dec, BPPARAM = bpp)
  
  set.seed(1000000)
  sce <- runUMAP(sce, dimred="PCA", BPPARAM = bpp)
  
  g <- buildSNNGraph(sce, k=30, use.dimred = 'PCA')
  clust <- igraph::cluster_walktrap(g)$membership
  colLabels(sce) <- factor(clust)
  
  plotUMAP(sce, colour_by="label")
  
  saveRDS(sce, file.path(data_dir, "human-pbmc", "pbmc.rds"))
  bpstop(bpp)
}

sce <- readRDS(file.path(data_dir, "human-pbmc", "pbmc.rds"))
cb_cluster_list <- colnames(sce)
names(cb_cluster_list) <- colLabels(sce)
cb_cluster_list <- split(cb_cluster_list, names(cb_cluster_list))
```

UMAP showing the assign clusters for this dataset

```{r}
plotUMAP(sce, colour_by="label")
```

## Quantify known editing sites in single cell data

```{r}
if(!file.exists(file.path(data_dir, "human-pbmc", "pbmc_pileups.rds"))){
  bamfn <- file.path(data_dir, "human-pbmc", "pbmc_subsampled_cbsorted.bam")
  bedfn <- file.path(data_dir, "rediportal_hg38.bed")
  # ~ 10 sec.
  build_tag_index(bamfn)
  dir.create(file.path(data_dir, "human-pbmc", "cluster-bams"), showWarnings = FALSE)
  # could do this in parallel using BiocParallel
  res <- list()
  for(i in seq_along(cb_cluster_list)){
    message("working on: ", names(cb_cluster_list)[i])
    cbs <- cb_cluster_list[[i]]
    cluster_bam <- get_cell_bam(bamfn, 
                                barcodes = cbs, 
                                outbam = file.path(data_dir, "human-pbmc",
                                                   "cluster-bams", 
                                                   paste0("cluster-", i, ".bam")))
    res[[i]] <- get_pileup(cluster_bam,
             fafile = file.path(data_dir,   
                                "Homo_sapiens.GRCh38.dna.primary_assembly.UCSC.fa"),
             bedfile = bedfn,
             min_reads = 1)
  }
  saveRDS(res, file.path(data_dir, "human-pbmc", "pbmc_pileups.rds"))
}

res <- readRDS(file.path(data_dir, "human-pbmc", "pbmc_pileups.rds"))
names(res) <- names(cb_cluster_list)
```

## Convert to SE object and do some filtering

```{r}
se <- create_se(res)

# replace missing values with 0, for now
assay(se, "nA")[is.na(assay(se, "nA"))] <- 0
assay(se, "nG")[is.na(assay(se, "nG"))] <- 0

# subset to sites with at least 10 A or G reads in at least 1 cluster
se_filtered <- se[rowSums(assay(se, "nA") + assay(se, "nG") >= 10) > 0, ]

# compute editing frequencies (A -> G only)
assay(se_filtered, "edit_freq") <- assay(se_filtered, "nG") / (assay(se_filtered, "nA") + assay(se_filtered, "nG"))
assay(se_filtered, "edit_freq")[is.na(assay(se_filtered, "edit_freq"))] <- 0

# subset to sites with at least 1 cluster with > 0.1 edit freq
se_filtered <- se_filtered[rowSums(assay(se_filtered, "edit_freq") > 0.1) > 0, ]
```

```{r}
library(ComplexHeatmap)
Heatmap(assay(se_filtered, "edit_freq"),
        name = "Edit-Frequency", 
        show_row_names = FALSE,
        show_row_dend = FALSE)
```

```{r}
library(data.table)
known_edit_info <- fread(file.path(data_dir, "TABLE1_hg38.txt.gz"), data.table = FALSE) 
known_edit_info$edit_id <- paste0(known_edit_info$Region, ":", known_edit_info$Position)

rrange <- rowRanges(se_filtered)
mcols(rowRanges(se_filtered))$edit_id <- paste0(seqnames(rrange), ":", start(rrange))
to_keep <- match(mcols(rowRanges(se_filtered))$edit_id, known_edit_info$edit_id )
edit_site_mdata <- known_edit_info[to_keep, ]
rownames(edit_site_mdata) <- rownames(se_filtered)
rowData(se_filtered) <- edit_site_mdata

rm(known_edit_info); tmp <- gc()
```

## Get data into SCE

For now just storing in the colData, could store as a separate altExp similar to CITE-seq.  

```{r}
cell_to_cluster <- data.frame(row.names = colnames(sce), sce$label)
edit_freq <- assay(se_filtered, "edit_freq")
rownames(edit_freq) <- rowData(se_filtered)$edit_id
# move to editing site x cluster
edit_freq <- t(edit_freq)
colData(sce) <- cbind(colData(sce), edit_freq[cell_to_cluster$sce.label, ])
```

## Select sites with editing in at least 5 clusters

```{r}
edit_freq <- t(edit_freq)
to_examine <- rownames(edit_freq[rowSums(edit_freq > 0) > 4, ])

edit_info <- rowData(se_filtered)[rowData(se_filtered)$edit_id %in% to_examine, c("edit_id", "repeat", "Func.refGene", "ExonicFunc.refGene")]
as.data.frame(edit_info)
```

The number of sites with editing frequency > 0 in at least 5 clusters `r  length(to_examine)`.


## Editing site plots {.tabset}

```{r, results = 'asis'}
max_to_plot <- min(50, nrow(edit_info))
library(purrr)
to_plot <- edit_info$edit_id
names(to_plot) <- edit_info$Func.refGene
iwalk(to_plot[1:max_to_plot], ~{
  if(.y %in% rownames(sce)){
    cat("### ", .y, " ", .x, '  \n')
    p1 <- plotUMAP(sce, colour_by = .y) + labs(subtitle = paste0(.y, "- expression"))
    p2 <- plotUMAP(sce, colour_by= .x) + labs(subtitle = paste0(.y, "- editing"))
    print(cowplot::plot_grid(p1, p2))
    cat("\n\n---\n\n<br>\n\n<br>\n\n")
  }
})

```


